qcengine/programs/entos.py

```diff
@@ -0,0 +1,238 @@
+            raise ModuleNotFoundError("""Python module jinja2 not found. Solve by installing it: `conda install 
+            jinja2` or `pip install jinja2`""")
+
+        # Write the geom xyz file with unit au
+        xyz_file = input_model.molecule.to_string(dtype='xyz', units='Angstrom')
+
+        # Create input dictionary
+        if template is None:
+            input_dict = {}
+            if input_model.driver == 'energy':
+                input_dict = {'dft': {
+                                'structure': {
+                                  'file': 'geometry.xyz'
+                                  },
+                                'xc': input_model.model.method,
+                                'ao': input_model.model.basis.upper(),
+                                'df_basis': input_model.keywords["df_basis"].upper(),
+                                'charge': input_model.molecule.molecular_charge
+                                },
+                              'print': {
+                                'results': True
+                                }
+                              }
+
+            # Write gradient call if asked for
+            if input_model.driver == 'gradient':
+                input_dict = {'gradient': {
+                                'structure': {
+                                  'file': 'geometry.xyz'
+                                  },
+                                'dft': {
+                                  'xc': input_model.model.method,
+                                  'ao': input_model.model.basis,
+                                  'df_basis': input_model.keywords["df_basis"].upper(),
+                                  'charge': input_model.molecule.molecular_charge
+                                   }
+                                 },
+                              'print': {
+                                'results': True
+                                }
+                              }
+            # Write input file
+            input_file = self.write_input_recursive(input_dict)
+            input_file = "\n".join(input_file)
+        else:
+            import jinja2
+            # Some of the potential different template options
+            # (A) ordinary build_input (need to define a base template)
+            # (B) user wants to add stuff after normal template (A)
+            # (C) user knows their domain language (doesn't use any QCSchema quantities)
+
+            # # Build dictionary for substitute
+            # sub_dict = {
+            #     "method": input_model.model.method,
+            #     "basis": input_model.model.basis,
+            #     "df_basis": input_model.keywords["df_basis"].upper(),
+            #     "charge": input_model.molecule.molecular_charge
+            # }
+
+            # Perform substitution to create input file
+            str_template = jinja2.Template(template)
```

Comment:
Something I was curious about is if we truly need jinja here, it seems our template are simple enough that they can be handled via more canonical means like finding `{{molecule}}` fields. I think my original suggestion may be been slightly optimistic towards the range of capabilities we could code in.

Reply:
I like the idea of having a template and then being able to do a dictionary substitute. I know the `string.Template` module offers some support of this, but it is very limited. I think we can provide the range of capabilities we discussed with jinja2, but it'll take some tinkering to be sure.

Reply:
Absolutely, jinja will totally cover our use case. But the question is do we need it? `string.Template` provides fairly reaching functionality.

Or even writing our own using `{{method}}` with double bracket replacement.

Reply:
I suppose what I didn't like about `string.Template` is that I couldn't find a way for it to tell me what fields are present which would then need to be filled. I'd like that functionality so we can be flexible in what information is required from the user in the `input.json`. It's possible that we could write our own but is it more trouble than it's worth? Is this mostly so we wouldn't need to have QCEngine be dependent on `jinja2` for something that is probably fairly simple?

Reply:
Pretty much, we try to keep the dependancies as small as possible for our lower level libraries and code small extra bits where needed.

Reply:
Okay so what is the way forward on this? @dgasmith I can switch over to using `string.Template` for now to keep the functionality I've added and to remove the jinja dependency, but I don't think this PR should include writing our own version of templating.

Reply:
I would vote to delay the jinja integration unless absolutely necessary. Lets work through this in the slack channel to figure out what we need.